### PR TITLE
Health Check For ACR

### PR DESCRIPTION
**Pull request by sowu880**

### CODE REVIEW

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        public AzureContainerRegistryHealthChecker(
+            IOptions<SchemaConfiguration> schemaConfiguration,
+            IContainerRegistryTokenProvider containerRegistryTokenProvider,
+            ILogger<AzureContainerRegistryHealthChecker> logger)
+            : base(HealthCheckTypes.AzureContainerRegistryCanRead, false, logger)
+        {
+            EnsureArg.IsNotNull(schemaConfiguration, nameof(schemaConfiguration));
+            _containerRegistryTokenProvider = EnsureArg.IsNotNull(containerRegistryTokenProvider, nameof(containerRegistryTokenProvider));
+
+            _schemaConfiguration = schemaConfiguration.Value;
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs:36]: Maybe we can only save the **SchemaImageReference** here? seems no other attributes be used.

Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs:36]: Or the converted **ImageInfo**

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                var accessToken = await _containerRegistryTokenProvider.GetTokenAsync(imageInfo.Registry, cancellationToken);
+                var acrClient = new AzureContainerRegistryClient(imageInfo.Registry, new AcrClientCredentials(accessToken));
+
+                // Ensure we can read from acr.
+                var result = await acrClient.Manifests.GetAsync(imageInfo.ImageName, imageInfo.Label, MediatypeV2Manifest, cancellationToken);
+            }
+            catch (Exception e)
+            {
+                healthCheckResult.Status = HealthCheckStatus.UNHEALTHY;
+                healthCheckResult.ErrorMessage = $"Read from acr failed. {e.Message}";
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs:56]: Do we need to log the information here?

Comment (sowu880) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs:56]: Will log outside. ( In baseHealthChecker.cs)

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/HealthCheckerRegistrationExtension.cs [hunk 3]
```diff
@@ -19,6 +23,23 @@ public static class HealthCheckerRegistrationExtension
...
             services.AddSingleton<IHealthChecker, AzureBlobStorageHealthChecker>();
 
+            try
+            {
+                var schemaConfiguration = services
+                    .BuildServiceProvider()
+                    .GetRequiredService<IOptions<SchemaConfiguration>>()
+                    .Value;
+
+                if (schemaConfiguration.EnableCustomizedSchema)
```
[the remark this answers is not in the document]
Comment (sowu880) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/HealthCheckerRegistrationExtension.cs]: Thanks, updated it.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs (hunk 1 above)
Comment (sowu880) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs]: Thanks, updated it.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            var healthCheckResult = new HealthCheckResult(HealthCheckTypes.AzureContainerRegistryCanRead, false);
+
+            try
+            {
+                var imageInfo = ImageInfo.CreateFromImageReference(_imageReference);
+                var accessToken = await _containerRegistryTokenProvider.GetTokenAsync(imageInfo.Registry, cancellationToken);
+                var acrClient = new AzureContainerRegistryClient(imageInfo.Registry, new AcrClientCredentials(accessToken));
+
+                // Ensure we can read from acr.
+                var result = await acrClient.Manifests.GetAsync(imageInfo.ImageName, imageInfo.Label, MediatypeV2Manifest, cancellationToken);
```
Comment (tongwu-msft) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs:50]: Need to check empty or null? If not, we can use "_" instead.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/HealthCheckerRegistrationExtension.cs [hunk 5]
```diff
@@ -19,6 +23,24 @@ public static class HealthCheckerRegistrationExtension
...
+            try
+            {
+                schemaConfiguration = services
+                    .BuildServiceProvider()
+                    .GetRequiredService<IOptions<SchemaConfiguration>>()
+                    .Value;
+            }
+            catch (Exception ex)
+            {
+                throw new ConfigurationErrorException("Failed to parse schema configuration", ex);
```
Comment (tongwu-msft) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/HealthCheckerRegistrationExtension.cs:36]: We parse the configuration before using it? Or it is a standard IOptions config?

Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/HealthCheckerRegistrationExtension.cs:36]: Feel this already be validated when initialization

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs (hunk 4 above)
Comment (sowu880) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/Checkers/AzureContainerRegistryHealthChecker.cs]: Updated it thanks.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/HealthCheckerRegistrationExtension.cs (hunk 5 above)
Comment (sowu880) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.HealthCheck/HealthCheckerRegistrationExtension.cs:36]: Configuration has been parsed and validated before it.

**Review by tongwu-msft — APPROVED** [review]

**Pull request merged**